lib/supavisor/tenants.ex

```diff
@@ -269,32 +243,25 @@ defmodule Supavisor.Tenants do
     User.changeset(user, attrs)
   end
 
-  @spec build_user_query(String.t(), String.t() | nil, String.t() | nil) :: Ecto.Queryable.t()
+  @spec build_user_query(String.t(), String.t() | nil, String.t() | nil) ::
+          Ecto.Queryable.t()
   defp build_user_query(user, external_id, sni_hostname) do
-    if sni_hostname do
-      query =
-        from(t in Tenant,
-          join: u in User,
-          on: u.tenant_external_id == t.external_id,
-          where: t.sni_hostname == ^sni_hostname,
-          select: {u, t}
-        )
-    else
-      query =
-        from(u in User,
-          join: t in Tenant,
-          on: u.tenant_external_id == t.external_id,
-          where: u.tenant_external_id == ^external_id,
-          select: {u, t}
-        )
-
-      if user do
-        from(u in query,
-          where: u.db_user_alias == ^user
-        )
-      else
-        query
-      end
-    end
+    from(u in User,
+      join: t in Tenant,
+      on: u.tenant_external_id == t.external_id,
+      where:
+        (u.db_user_alias == ^user and t.require_user == true) or
+          t.require_user == false,
+      select: {u, t}
+    )
+    |> where(^with_tenant(external_id, sni_hostname))
+  end
+
+  defp with_tenant(nil, sni_hostname) do
+    dynamic([_, t], t.sni_hostname == ^sni_hostname)
+  end
+
+  defp with_tenant(external_id, _) do
+    dynamic([_, t], t.external_id == ^external_id)
```

Comment:
I believe this also works but they all work the same. :)

```suggestion
    [sni_hostname: sni_hostname]
  end

  defp with_tenant(external_id, _) do
    [external_id: external_id]
```